File: sample+utils/mgpsd-via-mmap.py
#!/usr/bin/env python

import  struct
import mmap
import time

file = open("/tmp/mgpstate", "r+")

data = mmap.mmap(file.fileno(), 244)

print "lat,lon"
while True:
    gpstate = struct.unpack("21i80h",data)

    print "lat=",float(gpstate[0])/1000000
    print "lon=",float(gpstate[1])/1000000
    print "alt=",float(gpstate[2])/1000

    print "pdop=",float(gpstate[3])/1000
    print "hdop=",float(gpstate[4])/1000
    print "vdop=",float(gpstate[5])/1000

    print "spd=",float(gpstate[6])/1000
    print "trk=",float(gpstate[7])/1000

    print gpstate[10],"/", gpstate[9], "/",gpstate[8]
    print gpstate[11],":", gpstate[12], ":",gpstate[13], ".", str(gpstate[14]).zfill(3)

    print "lock=",gpstate[15]
    print "fix=", gpstate[16]

# BT status    print gpstate[17], gpstate[18]

    sats = gpstate[19]
    print "sats=", sats," used=", gpstate[20]

    for i in range(sats):
        j=21+i*4
        print gpstate[j],gpstate[j+1],gpstate[j+2],gpstate[j+3]

    time.sleep(0.3)
    
'''
struct satinfo {
    unsigned char num, el, az, sn;
};

struct gpsstate {
    int llat, llon, alt;
    int pdop, hdop, vdop;
    int gspd, gtrk;
    int yr, mo, dy, hr, mn, sc, scth;
    int lock, fix;
    int gpsfd, obdfd;
    int nsats, nused;
    struct satinfo sats[20];
};
'''

                                                

File: sample+utils/mgpsd-via-mmap.c
#include <fcntl.h>
#include <sys/mman.h>
#include <stdio.h>
#include <unistd.h>
#include "minigpsdmmap.h"

int main()
{
    int file;
    struct gpsstate *data;

    file = open("/tmp/mgpstate", O_RDONLY);
    data = mmap(0, sizeof(struct gpsstate), PROT_READ, MAP_SHARED, file, 0);

    for (;;) {
        sleep(1);

        printf("lat=%f\n", (double) data->llat / 1000000.0);
        printf("lon=%f\n", (double) data->llon / 1000000.0);
        printf("alt=%f\n", (double) data->alt / 1000.0);

        printf("pdop=%f\n", (double) data->pdop / 1000.0);
        printf("hdop=%f\n", (double) data->hdop / 1000.0);
        printf("vdop=%f\n", (double) data->vdop / 1000.0);

        printf("spd=%f\n", (double) data->gspd / 1000.0);
        printf("trk=%f\n", (double) data->gtrk / 1000.0);

        printf("%d/%d/%02d\n", data->mo, data->dy, data->yr);
        printf("%02d:%02d:%02d.%03d\n", data->hr, data->mn, data->sc, data->scth);

        printf("lock=%d\n", data->lock);
        printf("fix=%d\n", data->fix);

        // BT status    printf( data->gpsfd, data->obdfd

        int sats = data->nsats;
        printf("sats=%d,", sats);
        printf(" used=%d\n", data->nused );
	printf("U Nu El Azm SN\n" );
        int i;
        for (i = 0; i < sats; i++)
            printf("%d %2d %2d %3d %2d\n", data->sats[i].num < 0, data->sats[i].num, data->sats[i].el, data->sats[i].az,
		   data->sats[i].sn);
    }
}
